patches/tModLoader/Terraria/ModLoader/ModNet.cs

```diff
@@ -227,12 +229,56 @@ internal static bool SyncClientMods(BinaryReader reader, out bool needsReload)
 
 			LocalMod matching = modFiles.FirstOrDefault(mod => header.Matches(mod.modFile));
 			if (matching != null) {
-				matching.Enabled = true;
+				if(clientMod != null) {
+					// Mod is enabled, but wrong version enabled
+					if(clientMod.Version > header.version) {
+						// TODO: Localize these messages once finalized
+						reloadRequiredExplanationEntries.Add(new ReloadRequiredExplanation(2, header.name, matching, $"[c/FFFACD:Switch] to v{header.version}\n(Temporarily downgrade from v{clientMod.Version})"));
+					}
+					else {
+						reloadRequiredExplanationEntries.Add(new ReloadRequiredExplanation(2, header.name, matching, $"[c/FFFACD:Switch] to v{header.version}\n(Temporarily upgrade from v{clientMod.Version})"));
+					}
+				}
+				else {
+					toEnable.Add(matching.Name);
+					reloadRequiredExplanationEntries.Add(new ReloadRequiredExplanation(3, header.name, matching, $"[c/98FB98:Enable]")); 
+				}
+
 				continue;
 			}
 
-			if (downloadModsFromServers && (header.signed || !onlyDownloadSignedMods))
+			if (downloadModsFromServers && (header.signed || !onlyDownloadSignedMods)) {
 				downloadQueue.Enqueue(header);
+				if (modFiles.FirstOrDefault(mod => mod.Name == header.name && mod.Version == header.version) is LocalMod localMod) {
```

Comment:
Didn't have time to parse this entire if statement, but it's definitely quite a bit of work to parse. Perhaps there's a way to factor or flatten it so the conditions are more clear